File: Reqnroll/Chapter13/13.4-EntitySelectorPattern/After/WIMP.Specs/Support/CustomParameterTypes.cs
using Reqnroll;

namespace WIMP.Specs.Support;

[Binding]
public class CustomParameterTypes(OrderingContext orderingContext)
{
    [StepArgumentTransformation("the earliest order received", Name = "order")]
    public int ConvertEarliestOrder()
    {
        return orderingContext.PlacedOrders
            .OrderBy(o => o.PlacingTime)
            .FirstOrDefault()?.OrderNo ??
            throw new InvalidOperationException("No orders available");
    }
}

File: CucumberJVM/Chapter13/13.4-EntitySelectorPattern/After/src/test/java/com/wimp/app/specs/support/CustomParameterTypes.java
package com.wimp.app.specs.support;

import com.wimp.app.models.*;
import io.cucumber.java.ParameterType;

import java.util.Comparator;
import java.util.Optional;

public class CustomParameterTypes {

    private final OrderingContext orderingContext;

    public CustomParameterTypes(OrderingContext orderingContext) {
        this.orderingContext = orderingContext;
    }

    @ParameterType("the order|the placed order|the earliest order received")
    public int order(String value) {
        if (value.matches("the order|the placed order")){
            Optional.ofNullable(orderingContext.getCurrentOrderNo())
                .orElseThrow(() -> new RuntimeException("No current order"));
        }
        if (value.matches("the earliest order received")){
            return orderingContext.getPlacedOrders().stream()
                .min(Comparator.comparing(Order::getPlacingTime))
                .map(Order::getOrderNo)
                .orElseThrow(() -> new RuntimeException("No orders available"));
        }
        throw new RuntimeException("Invalid order value: " + value);
    }
}
